### PR TITLE
Add Qt-main recipe

### PR DESCRIPTION
**Pull request by andfoy**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

This PR makes part of the Qt split that was discussed last week during the conda-forge core meeting.

### CODE REVIEW

**Comment by andfoy** [comment]
@conda-forge/staged-recipes is it possible to run recipes on CentOS 7 here?

**Comment by isuruf** [comment]
Please send this PR to `qt-feedstock` first so that we can review it and compare against the current recipe.

**Comment by andfoy** [comment]
The test-run PR is available at: https://github.com/conda-forge/qt-feedstock/pull/198

**Comment by andfoy** [comment]
@isuruf, do you know why the linter is complaining? I don't see any `qt-main-feedstock` in the organization

File: recipes/qt-main/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,241 @@
+{% set name = "qt" %}
```
Comment (isuruf) [inline recipes/qt-main/meta.yaml:1]:
```suggestion
{% set name = "qt-main" %}
```

File: recipes/qt-main/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    - libclang
+    - xorg-libxfixes                     # [linux]
+    - libcups                            # [linux]
+  run:
+    - {{ pin_compatible("nss") }}        # [unix]
+    - {{ pin_compatible("nspr") }}       # [unix]
+    - libxcb                             # [linux]
+    - {{ pin_compatible("libclang") }}
+  run_constrained:                                            # [osx]
+    - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.14") }}  # [osx]
```
Comment (isuruf) [inline recipes/qt-main/meta.yaml:170]:
```suggestion
  run_constrained:
    - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.14") }}  # [osx]
    - qt {{ version }}
```

Comment (andfoy) [reply recipes/qt-main/meta.yaml:170]: Isn't it `qt-main` there?

**Comment by andfoy** [comment]
@isuruf, do we need tests to pass here?

**Comment by andfoy** [comment]
@isuruf, this recipe was not designed with CentOS 6 in mind, and it won't work on it. Also we require clang 12 on Mac, which cannot be set up here unless the version is hardcoded in `conda_build_config`. Is it possible to merge it as it was approved?

**Comment by isuruf** [comment]
> Also we require clang 12 on Mac, which cannot be set up here unless the version is hardcoded in conda_build_config

I don't understand this.

> @isuruf, this recipe was not designed with CentOS 6 in mind, and it won't work on it

Both CentOS 6 and 7 is tested here. So the CentOS 7 job should work.

**Comment by andfoy** [comment]
> I don't understand this.

For some reason, the dependency on libclang is version 13 and the compilers (via `clang_osx`) is 11, this error was fixed on the feedstock by constraining the libclang version to the same version that `c/cxx_compiler_version` has

**Comment by isuruf** [comment]
> For some reason, the dependency on libclang is version 13 and the compilers (via clang_osx) is 11, this error was fixed on the feedstock by constraining the libclang version to the same version that c/cxx_compiler_version has

What's different from the feedstock and staged-recipes?

**Comment by andfoy** [comment]
https://github.com/conda-forge/qt-feedstock/blob/2fe1037a06bfe17b6b2a3ab48a8ce56c752d949b/.ci_support/osx_64_.yaml

**Comment by isuruf** [comment]
You were editing that manually?

**Comment by andfoy** [comment]
Yes, I did edit that manually

**Comment by isuruf** [comment]
That's not allowed. Those files should not be touched at all.

**Comment by andfoy** [comment]
What should be the correct way?

**Comment by ngam** [comment]
> What should be the correct way?

@andfoy, here usually:
https://github.com/andfoy/staged-recipes/blob/add_qt_main/recipes/qt-main/conda_build_config.yaml

Add something like:
```
c_compiler_version:    # [osx]
- '12'                 # [osx]
cxx_compiler_version:  # [osx]
- '12'                 # [osx]
```

File: recipes/qt-main/conda_build_config.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+MACOSX_DEPLOYMENT_TARGET:  # [osx and x86_64]
+  - '10.13'                # [osx and x86_64]
+MACOSX_SDK_VERSION:        # [osx and x86_64]
+  - '10.14'                # [osx and x86_64]
```
Comment (ngam) [inline recipes/qt-main/conda_build_config.yaml:4]: add

```
c_compiler_version:    # [osx]
- '12'                 # [osx]
cxx_compiler_version:  # [osx]
- '12'                 # [osx]
```

**Review by ngam — CHANGES_REQUESTED** [review]

**Comment by andfoy** [comment]
@ngam , I was aware of this before. The error was caused by an internal repodata constraint that was fixed in

**Comment by ngam** [comment]
> @ngam , I was aware of this before. The error was caused by an internal repodata constraint that was fixed in [conda-forge/conda-forge-repodata-patches-feedstock#195](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/195)

Oh, okay good 👍

**Comment by Tobias-Fischer** [comment]
Latest error on linux_64_centos7 for reference:
```
2021-12-06T19:47:12.1233826Z conda_build.exceptions.OverLinkingError: overlinking check failed 
2021-12-06T19:47:12.1242458Z ['  
ERROR (qt-main,lib/libQt5EglFsKmsSupport.so.5.15.2): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  
ERROR (qt-main,plugins/mediaservice/libgstmediacapture.so): .. but ['alsa-lib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", "  
ERROR (qt-main,plugins/audio/libqtaudio_alsa.so): .. but ['alsa-lib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  
ERROR (qt-main,plugins/egldeviceintegrations/libqeglfs-x11-integration.so): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  
ERROR (qt-main,lib/libQt5MultimediaGstTools.so.5.15.2): .. but ['alsa-lib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", "  
ERROR (qt-main,plugins/mediaservice/libgstaudiodecoder.so): .. but ['alsa-lib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", "  
ERROR (qt-main,plugins/mediaservice/libgstcamerabin.so): .. but ['alsa-lib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", "  
ERROR (qt-main,lib/libQt5Network.so.5.15.2): .. but ['krb5'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", "  
ERROR (qt-main,bin/rcc): .. but ['zstd'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", " 
ERROR (qt-main,plugins/mediaservice/libgstmediaplayer.so): .. but ['alsa-lib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  
ERROR (qt-main,lib/libQt5EglFSDeviceIntegration.so.5.15.2): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  
ERROR (qt-main,plugins/xcbglintegrations/libqxcb-egl-integration.so): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  
ERROR (qt-main,plugins/egldeviceintegrations/libqeglfs-emu-integration.so): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  
ERROR (qt-main,plugins/platforms/libqminimalegl.so): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  
ERROR (qt-main,lib/libQt5Core.so.5.15.2): .. but ['zstd'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  
ERROR (qt-main,plugins/egldeviceintegrations/libqeglfs-kms-egldevice-integration.so): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  
ERROR (qt-main,plugins/platforms/libqeglfs.so): $RPATH/libEGL.so.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
```

So a few issues around `$RPATH/libEGL.so.1`, `zstd`, `krb5` and `alsa-lib`.

The libEGL issue can probably be resolved by adding mesa-libEGL to yum_requirements.txt. I'm not sure about the others though ..

**Comment by isuruf** [comment]
Thanks

**Comment by andfoy** [comment]
Thanks a lot for the help with the reviewing process

**Pull request merged**